pageserver/src/page_cache.rs

```diff
@@ -771,14 +883,9 @@ impl PageCache {
                 let mut inner = match slot.inner.try_write() {
                     Ok(inner) => inner,
                     Err(_err) => {
-                        // If we have looped through the whole buffer pool 10 times
-                        // and still haven't found a victim buffer, something's wrong.
-                        // Maybe all the buffers were in locked. That could happen in
-                        // theory, if you have more threads holding buffers locked than
-                        // there are buffers in the pool. In practice, with a reasonably
-                        // large buffer pool it really shouldn't happen.
                         if iters > iter_limit {
-                            anyhow::bail!("exceeded evict iter limit");
+                            // we take the _permit_witness as proof
+                            unreachable!("we hold a slot permit, there mut be one not-locked slot");
```

Comment:
`unreachable` makes me a little nervous, I'd still prefer it to be `bail`. Even with the semaphore, there's no hard guarantee that you will find a slot with any particular number of iterations: other threads might race ahead and acquire and release pins just as you're scanning the array. Highly unlikely but not impossible.

A more likely cause is that we have a bug somewhere. If that happens, do we want to panic? Who knows what went wrong, but it feels likely that it would be recoverable, the caller ought to handle an error better than a panic..

(typo: `mut` -> `must`)

Reply:
> Even with the semaphore, there's no hard guarantee that you will find a slot with any particular number of iterations: other threads might race ahead and acquire and release pins just as you're scanning the array. Highly unlikely but not impossible.

Hm, we don't allow more than nslots outstanding permits and only search for slots while we hold a permit. Doesn't this guarantee that we find a free slot after within `iter_limit := MAX_USAGE_COUNT * nslots` iterations?

That was kind of the whole point of this PR.

Reply:
> Doesn't this guarantee that we find a free slot after within `iter_limit := MAX_USAGE_COUNT * nslots` iterations?

Definitely, but refactors might change that, and there might also be bugs or mistakes in our understanding. Compare with how a refactor that didn't touch the page cache file brought us into the situation of hitting this bail instance. While I do think that this PR makes the cases way way more unlikely, to the point that I don't think it will be hit, I think that planning for the case that one is wrong isn't a bad idea.

Therefore, I don't really think that having a panic is a good idea, at least for now while we don't have HA secondaries to fail over to. Once HA is there, I think a panic might make more sense, for reasons stated below.

To give arguments for panicing, the pageserver is a shared resource and if this error is reached, then basic assumptions/invariants are violated. Invariants are what sustain tenant separation too, so it might actually be better to have downtime than running the risk of the invariant violation being so serious that tenant data is leaked. It's the old "should we continue running in an invalid state, or should we never ever try to recover from logic bugs even when it would just be a retry" question.

Reply:
> > Even with the semaphore, there's no hard guarantee that you will find a slot with any particular number of iterations: other threads might race ahead and acquire and release pins just as you're scanning the array. Highly unlikely but not impossible.
> 
> Hm, we don't allow more than nslots outstanding permits and only search for slots while we hold a permit. Doesn't this guarantee that we find a free slot after within `iter_limit := MAX_USAGE_COUNT * nslots` iterations?

No. Imagine that nslots is 2, and as starting point, usage_count==1 on all slots. There are two threads running concurrently, A and B. A has just acquired the permit from the semaphore.

A: Look at slot 1. Its usage_count == 1, so decrement it to zero, and continue the search
B: Acquire permit.
B: Look at slot 2, decrement its usage_count to zero and continue the search
B: Look at slot 1. Its usage_count is zero, so pin it and bump up its usage_count to 1.
B: Release pin and permit again
B: Acquire permit.
B: Look at slot 2. Its usage_count is zero, so pin it and up up its usage_count to 1.
B: Release pin and permit again

Now we're back in the starting situation that both slots have usage_count 2, but A has so now been through one iteration of the find_victim() loop. You can repeat the above indefinitely if you're (un)lucky, and on each iteration, A's iteration count increases by one.

This is highly unlikely to happen in practice or indefinitely, especially if nslots is a reasonably high number. But in theory it's possible, there's no guarantee that you find a victim after any particular number of iterations.

Reply:
Hm, right, failed to see that.
The semaphore is fair, but the victim search still happens in parallel.
This also means that the starvation issue isn't fully solved with this PR, just reduced a lot because once we're searching for a victim, we're not yielding to the tokio scheduler.

Reply:
I guess the only solution to ensure we find a victim is to serialize `find_victim``. Which by gut feeling is probably worse.

I'll do the anyhow::bail then.

Reply:
I guess the alternative is to remove the `iter_limit` and yield to the runtime?

Reply:
yielding might be counter productive because then we give the 'B' task and tasks like it more opportunities to increase usage_count. The condition for us spinning many times in the loop is other tasks being successful more quickly than us.

it's not preventable though that the more business there is in the page cache, the more likely we are going to spin. we could somehow counteract that by adding a flexible free area in the page cache, as in, making the permit limit smaller than the actual number of pages in the cache. how large the deviation of the two numbers should be probably depends on the max number of parallelism we have (depends on number of cores and if we await also on the number of tasks). But such stuff we should probably only add if we see this error regularly.

regarding the original motivation to add a yield here, does pages being freed help us in this situation? we know that there is a page somewhere that's free, but probably the more free pages, the less likely this situation occurs. so I'd say yes.

I think the best pro yield argument is that it would follow the general rule that we should yield regularly if we are busy. Even if it means it takes longer as other tasks can bump the usage_count numbers.

Reply:
I mean, if we return an error here, we also yield to the executor, we just do it way more noisily (iirc the retry logic has async waits inside?). Running over the entire array 5 times should be within the target time between yields. The noise might make sense because the current assumption is that it's unlikely to happen and if it happens often we would like to be informed. We can also just log something and yield for some small limit, and error for higher limit to escalate the noise level. We can also just have the high noise level at the start and add the prior thing if the limit is hit often.